Ensure raft membership

trozet opened the pull request:
There are two scenarios where a stale member may be leftover in the Raft
    membership:
    
    1. OVN DB pod goes down, database is wiped, comes back up with same
    address but gets a new server ID in Raft
    
    2. DB node is removed from cluster and replaced with a new one with a
    different address
    
    For 1, each ovnkube-master can check its current ID and look for any
    leftover members with its same address.
    
    For 2, only ovnkube-master leader will compare the known addresses
    passed to it for other DB members, and kick out any stale members.
    
    Signed-off-by: Tim Rozet <trozet@redhat.com>

Billy99 commented on dist/images/ovnkube.sh:
```diff
@@ -290,13 +290,14 @@ check_ovn_daemonset_version() {
   ovn_sbdb_str=""
-  for i in ${!ovn_db_hosts[@]}; do
-    if [[ ${i} -ne 0 ]]; then
+  for i in "${ovn_db_hosts[@]}"; do
```
There are two more brackets on IPv4 addresses. Not sure if you care about those:
~Line 654:
  ovn-nbctl --inactivity-probe=0 set-connection p${transport}:${ovn_nb_port}:[${ovn_db_host}]
~Line 686
  ovn-sbctl --inactivity-probe=0 set-connection p${transport}:${ovn_sb_port}:[${ovn_db_host}]

trozet replied:
thanks I'll fix this

fedepaol commented on go-controller/pkg/ovn/master.go:
```diff
@@ -85,6 +86,8 @@ func (oc *Controller) Start(kClient kubernetes.Interface, nodeName string, wg *s
 				// run the End-to-end timestamp metric updater only on the
 				// active master node.
 				metrics.StartE2ETimeStampMetricUpdater(oc.stopChan, oc.ovnNBClient)
+				go oc.ensureClusterRaftMembership(util.OvnNbdbLocation)
```
Could it happen that we are not the leader anymore and `OnStoppedLeading` gets called? In that case, would it make sense to stop these?

trozet replied:
If leader stops it will os.exit so the goroutine will get killed anyway. If the master for some reason got stopped but was still leader oc.stopChan would close and that would also stop the goroutine.

fedepaol replied:
Got it, so one active and healthy leader cannot be replaced by a new one for any reason. Thanks!

fedepaol commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1056,3 +1061,128 @@ func (oc *Controller) deleteNodeChassis(nodeName string) error {
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+		knownMembers = strings.Split(config.OvnNorth.Address, ",")
```
This means that in order to add a new member, the configuration needs to be updated and the leader (or potentially, all the masters) needs to be restarted. right? Otherwise, as soon as it sees it, it will kick it out.

trozet replied:
Right. So in OCP, CNO will refresh the daemonset with the new addresses, and in CNO the ovnkube-master DS contains both ovnkube-master and the dbs. So when DS gets refreshed, everything will restart and I don't think this will be an issue.

But for upstream and to be extra safe...If somehow the new one came up and was kicked because the old members were not updated yet, then we should expect the affected db to exit and rejoin the cluster when he restarts. However I do not see any check in the db code today to ensure he is still part of the cluster. I'll look at adding that and report back.

trozet replied:
I think I resolved this by adding commit "Add health check to ensure Raft membership"

fedepaol reviewed (commented):
A couple of questions, mostly for my understanding

girishmg commented:
@cathy-zhou can you PTAL at this PR?

girishmg commented:
@trozet few questions:
- in real deployment, when is (1) is an issue? The OVN DB files are on the hostpath?
- also in real deployment, we should pin the OVN DB Pods to a set of K8s nodes so that the OVN DB IPs are same. this is akin to how etcd servers will be managed for K8s cluster

trozet commented:
/hold I have another commit that fixes how we configure raft membership in the first place when dbs come up

trozet commented:
/hold cancel

trozet commented:
> @trozet few questions:
> 
> * in real deployment, when is (1) is an issue? The OVN DB files are on the hostpath?

DB doesn't necessarily have to be removed on the host, could also be the DB is corrupted. Which is what we are seeing in:
https://bugzilla.redhat.com/show_bug.cgi?id=1837953

I think we identified the cause of the corruption, but in this case we should be able to recover.


> * also in real deployment, we should pin the OVN DB Pods to a set of K8s nodes so that the OVN DB IPs are same. this is akin to how etcd servers will be managed for K8s cluster

I don't think that is realistic. A customer could want to add another master node temporarily while he brings down his other one for a hardware upgrade. We should be able to add new ones, and remove old.

In either case this PR makes things more robust, by removing assumptions when db pods are started, and monitoring their health and raft membership more carefully.

cathy-zhou commented:
I have several basic questions. This change are based on several assumptions: 

- ovnkube-db-raft and ovnkube-master Pods are running on the same set of nodes. Is this always the case? This also include the need that the replica number of ovnkube-master deployment is the same as replica of ovnkube-db-raft statefulset.
- in order for (2) to work, ovnkube-db-raft pod on the ovnkube-master leader needs to be active and running, this may not always be the case either.

BTW, for (1), in case of ovn DB corruption, in order to recover from that, we'd still need the logic to rejoin the cluster, which is not there for ovnkube-db-0.

trozet commented:
> I have several basic questions. This change are based on several assumptions:
> 
> * ovnkube-db-raft and ovnkube-master Pods are running on the same set of nodes. Is this always the case? This also include the need that the replica number of ovnkube-master deployment is the same as replica of ovnkube-db-raft statefulset.

Correct. In our usage in OCP this is always the case since db is bundled with ovn-k8s master DS. In upstream here we have them split into different pods, but they always are on the same master nodes. Do you think this is a problem?

> * in order for (2) to work, ovnkube-db-raft pod on the ovnkube-master leader needs to be active and running, this may not always be the case either.

I added a check for the db to make sure it is part of the cluster, otherwise it exits. That should make sure it eventually comes back up so the ovn-k8s leader can do its thing.

> 
> BTW, for (1), in case of ovn DB corruption, in order to recover from that, we'd still need the logic to rejoin the cluster, which is not there for ovnkube-db-0.

So you may  have reviewed this before I added my other 2 commits. I refactored that code so that ovnkube-db-0 will rejoin the cluster.

trozet commented:
@cathy-zhou I should also mention there is going to be another PR to check the database itself and ensure it is not corrupt (and handle it if it is), but this PR's focus is simply handling raft membership.

trozet commented:
I'm going to move the checks to go on ovnkube-node and use a check to see if ovn db is on that node (similar to the metrics code)

trozet commented:
@girishmg @cathy-zhou I updated the golang part to be part of ovnkube-node and only check if dbs are on that node. PTAL

girishmg commented:
@trozet you have made lot of changes to the critical part of the system. Need some time to digest your changes :)

For example: In your last commit, you have made changes to check_health() which looks dangerous to me. What will happen if `check_health` fails on all of the three nodes around the same time? This PR will end up deleting the OVN NB and SB DB files on all the nodes. At scale, we have seen instances where `cluster/status` just fails because `ovn-appctl` timeouts maybe because the DB is legitimately busy. We don't want to delete the DB files in such instances.

girishmg commented on go-controller/pkg/node/ovn_db_mgmt.go:
```diff
@@ -0,0 +1,139 @@
+		return
+	}
+	r, _ := regexp.Compile(`Address: *((ssl|tcp):[?[a-z0-9.:]+]?)`)
+	matches := r.FindStringSubmatch(out)
```
There are easier ways to get the Server ID from cluster/status output, so  I don't know why we need to start from `Address: `. In the output below , the very first line is Server ID and there is a column for 'Server ID' itself.


```
]# ovs-appctl -t /var/run/openvswitch/ovnnb_db.ctl cluster/status OVN_Northbound
42db    <============== **SERVER ID**
Name: OVN_Northbound 
Cluster ID: d92f (d92fa139-71be-4beb-b361-0ec571d75965)
Server ID: 42db (42dbe4b3-9a94-4569-93fd-ee3dd17bdc12) <========= ** SERVER ID **
Address: tcp:10.0.16.14:6643
Status: cluster member
Role: follower
Term: 75
Leader: 71f8
Vote: 71f8

Election timer: 5000
Log: [87261, 88800]
Entries not yet committed: 0
Entries not yet applied: 0
Connections: ->7e36 ->71f8 <-7e36 <-71f8
Servers:
    7e36 (7e36 at tcp:10.0.16.29:6643)
    71f8 (71f8 at tcp:10.0.2.20:6643)
    42db (42db at tcp:10.0.16.14:6643) (self)
```

trozet replied:
In this particular line im getting the address. I get the server id above on line 50: `out, stderr, err := util.RunOVSDBTool("db-sid", db)`

girishmg replied:
So, you get the ServerID from the DB file directly using `ovsdb-tool db-sid` and then you get the ServerID from the running daemon, and you do compare.

I am talking about getting the ServerID from the running daemon. Why can't we just take the value from the 1st line in the `cluster/status` output.

trozet replied:
We can. I thought sb-id might be more authoritative.

girishmg commented on go-controller/pkg/node/ovn_db_mgmt.go:
```diff
@@ -0,0 +1,139 @@
+	r, _ = regexp.Compile("([a-z0-9]{4}) at " + addr)
+	members := r.FindAllStringSubmatch(out, -1)
+	for _, member := range members {
+		if len(member) < 2 {
```
the `len(members)` will be 1 since your regex is bound to `addr`? 

Looks like some of the code is from when it was running from the `ovnkube-master`?

trozet replied:
its a submatch so there should be 2 entries, the entire match, plus the submatch of the ID.

Which code are you talking about from ovnkube-master?

girishmg replied:
@trozet I am talking about `members` and not `member`. The `members` length will be always one since you are matching the output against the local node's address, right?

trozet replied:
@girishmg no that's not the case. The issue this PR is solving is when you bring back a node with a new db it will actually have 2 entries in the raft cluster with the same address: one for the previous ID and one for the new ID:

```
[root@ovn-control-plane2 ~]# ovn-appctl -t /var/run/ovn/ovnnb_db.ctl cluster/status OVN_Northbound
e60c
Name: OVN_Northbound
Cluster ID: bcb7 (bcb74ae8-ba91-47fa-afe8-f436829b079e)
Server ID: e60c (e60cff74-fa76-4bb4-96d6-4c740dd615f9)
Address: tcp:[172.18.0.5]:6643
Status: cluster member
Role: leader
Term: 1
Leader: self
Vote: self

Election timer: 1000
Log: [2, 185]
Entries not yet committed: 0
Entries not yet applied: 0
Connections: <-9d15 ->d265 ->9d15 <-13f8
Servers:
    9d15 (9d15 at tcp:[172.18.0.3]:6643) next_index=185 match_index=184
    e60c (e60c at tcp:[172.18.0.5]:6643) (self) next_index=2 match_index=184
    13f8 (13f8 at tcp:[172.18.0.4]:6643) next_index=185 match_index=184
    d265 (d265 at tcp:[172.18.0.4]:6643) next_index=185 match_index=182
```

girishmg commented on go-controller/pkg/node/ovn_db_mgmt.go:
```diff
@@ -0,0 +1,139 @@
+			if err != nil {
+				klog.Errorf("Error while kicking old Raft member: %s, for address: %s in db: %s,"+
+					"stderr: %v, error: %v", member[1], addr, db, stderr, err)
+			}
```
you cannot `kick` yourself out. You will have run `cluster/kick` from the other DB nodes. 

Please take a look at **How to Maintain a Clustered Database** at the link here https://docs.openvswitch.org/en/latest/ref/ovsdb.7/#how-to-maintain-a-clustered-database

trozet replied:
you can kick yourself out. I tested it

girishmg replied:
That is only if the node has not crashed or down from the cluster or some such. As I understand, `kick` is meant for remaining set of healthy nodes in a cluster to kick an unhealthy node.

Also, the cluster/kick will only work if the cluster is healthy. That is why, we need an observer (aka operator) that looks at the cluster from outside. Figures out what is wrong with the cluster as a whole -- which node is up and which node is down due to corrupted db or crashed db  and so on -- and then makes a informed decision.

Applying fixes in a localized manner will not work.

trozet replied:
this specific function is for kicking a previous id out of the cluster using the same address as this node's current address. There's no way for an outside operator to know which is the valid one. Only the local node knows which one is his current id.

trozet commented:
> @trozet
> 
> To identify RAFT issues and remediate some of those issues, we need a `ovsdb` operator that takes the operational aspects of all the 3-nodes and makes a decision.
> 
> For example: AFAIU, you cannot run "cluster/kick" on a node to kick itself out. The other nodes in the cluster should kick the node out. The link here https://docs.openvswitch.org/en/latest/ref/ovsdb.7/#how-to-maintain-a-clustered-database is very useful.

You can kick yourself out of the cluster. I was doing this to test out my patch.

trozet commented on go-controller/pkg/util/ovs.go:
```diff
@@ -47,8 +47,10 @@ const (
+	nbdbCtlSock     = "ovnnb_db.ctl"
+	sbdbCtlSock     = "ovnsb_db.ctl"
+	OvnNbdbLocation = "/etc/openvswitch/ovnnb_db.db"
+	OvnSbdbLocation = "/etc/openvswitch/ovnsb_db.db"
```
will fix

trozet commented on go-controller/pkg/node/ovn_db_mgmt.go:
```diff
@@ -0,0 +1,139 @@
+				klog.Infof("Node %s has a OVN DB pod, ensuring OVN Raft membership", n.name)
+				n.ensureLocalRaftServerID(db)
+				n.ensureClusterRaftMembership(db)
+			}
```
There's a couple options here. I could just make it only check 1 time (at startup) if this node has a db co-located. It would handle the case of adding a new node with a db, but would not handle if a db pod was removed or added to an existing node. Doesn't seem great.

@girishmg what if we make it into its own go binary and run it as part of the ovnkube-db container via ovnkube.sh? Is there a reason you think it needs to be a separate container? I wonder if then we could move a lot of the other bash into golang here as well...

girishmg replied:
Ideally, we would need a OVSDB operator to remediate things correctly. You cannot fix RAFT by running commands on a node. These commands will work only if the RAFT is healthy (has majority number of nodes healthy).

Yes, keeping all these changes in ovnkube.sh itself works. We don't need separate container. I would rather just keep all this in the script itself instead of a go binary. YMMW

trozet replied:
I think maybe we need to have a meet to discuss this stuff. I'm not sure I explained what this was fixing well enough, and I'd like to come to a consensus on where this stuff should run. I'll ping you on slack.

trozet commented:
> @trozet you have made lot of changes to the critical part of the system. Need some time to digest your changes :)
> 
> For example: In your last commit, you have made changes to check_health() which looks dangerous to me. What will happen if `check_health` fails on all of the three nodes around the same time? This PR will end up deleting the OVN NB and SB DB files on all the nodes. At scale, we have seen instances where `cluster/status` just fails because `ovn-appctl` timeouts maybe because the DB is legitimately busy. We don't want to delete the DB files in such instances.

I actually saw in KIND that sometimes we would timeout as well. @girishmg when I was adding that health check I debated making that check more specific, and I think with your feedback I probably should. We can check to see if the return code is 2, and the ouput is "Unknown". That will ensure its a real response and not a connection problem indicating the health is bad.

cathy-zhou commented on dist/images/ovndb-raft-functions.sh:
```diff
@@ -238,26 +274,53 @@ ovsdb-raft() {
-    fi
-    run_as_ovs_user_if_needed \
+  initial_raft_create=true
+  if [[ "${initialize}" == "true" ]]; then
```
ovsdb-server needs to be started when ovnkube-db-raft pod restarts (where $initialize is false).

trozet replied:
thanks. will fix.

cathy-zhou commented on dist/images/ovndb-raft-functions.sh:
```diff
@@ -272,12 +335,14 @@ ovsdb-raft() {
+      fi
     fi
     # set the connection and disable inactivity probe, this deletes the old connection if any
     set_connection ${db} ${port}
```
move set_connection into "if ${initial_raft_create}; then" bracket?

trozet replied:
In the comments it says:
# set_connection() will be called for ovnkube-db-0 pod when :
#    1. it is first started or
#    2. it restarts after the initial start has failed or
#    3. subsequent restarts during the lifetime of the pod

cathy-zhou replied:
yes, but now the condition has changed, if $initial_raft_create is true, it means connection has already been set to the cluster (it passed the db_part_of_cluster() check which does a list-connection).

girishmg commented:
> > @trozet
> > To identify RAFT issues and remediate some of those issues, we need a `ovsdb` operator that takes the operational aspects of all the 3-nodes and makes a decision.
> > For example: AFAIU, you cannot run "cluster/kick" on a node to kick itself out. The other nodes in the cluster should kick the node out. The link here https://docs.openvswitch.org/en/latest/ref/ovsdb.7/#how-to-maintain-a-clustered-database is very useful.
> 
> You can kick yourself out of the cluster. I was doing this to test out my patch.

please see this: https://github.com/ovn-org/ovn-kubernetes/pull/1637#discussion_r478572772

cathy-zhou commented on dist/images/ovnkube.sh:
```diff
@@ -290,13 +293,14 @@ check_ovn_daemonset_version() {
   ovn_sbdb_str=""
-  for i in ${!ovn_db_hosts[@]}; do
-    if [[ ${i} -ne 0 ]]; then
+  for i in "${ovn_db_hosts[@]}"; do
```
rename “i” to "ip" instead to make it clear? "i" usually refers to an interger.

trozet replied:
sure

cathy-zhou commented on dist/images/ovnkube.sh:
```diff
@@ -398,6 +402,24 @@ check_health() {
   esac
 
+  if $DB_CLUSTERED; then
+    if [[ ${1} == "ovnnb_db" ]]; then
```
so as long as ovsdb-server for NB/DB exits (E.g. coredump), we wipe out the DB file? This is quite intrusive.

trozet replied:
no this was only intended to wipe the db if we are no longer part of the cluster, to force rejoin with a clean slate, then the raft cluster woudl resync our db. However, I agree that this is probably too aggressive. Instead of just checking if the command doesn't return 0, we can check if the command returns 2 and "Unknown" which will signal to us that the node is no longer in a cluster.

cathy-zhou commented on go-controller/pkg/node/ovn_db_mgmt.go:
```diff
@@ -0,0 +1,139 @@
+		appCtl = util.RunOVNSBAppCtl
+	}
+
+	out, stderr, err := util.RunOVSDBTool("db-sid", db)
```
considering running "ovs-appctl -t /var/run/ovn/ovnnb_db.ctl cluster/sid OVN_Northbound" to get the current server id? Getting from running state is probably better?

trozet replied:
i guess this comes down to what is a better source of truth, the database or the memory state? I would argue the database should always be the source of truth.

cathy-zhou commented on go-controller/pkg/node/ovn_db_mgmt.go:
```diff
@@ -0,0 +1,139 @@
+				"name=ovnkube-db", n.name, false); dbOnNode {
+				klog.Infof("Node %s has a OVN DB pod, ensuring OVN Raft membership", n.name)
+				n.ensureLocalRaftServerID(db)
+				n.ensureClusterRaftMembership(db)
```
so every node Pod running ovnkube-raft nodes would do ensureClusterRaftMembership()? Is it better to only run it on the cluster leader?

trozet replied:
previously i was only running it on the ovn-kube master leader, but yeah I think it would make sense to only run it where DB leader currently resides.

girishmg commented:
> > * also in real deployment, we should pin the OVN DB Pods to a set of K8s nodes so that the OVN DB IPs are same. this is akin to how etcd servers will be managed for K8s cluster
> 
> I don't think that is realistic. A customer could want to add another master node temporarily while he brings down his other one for a hardware upgrade (without having to change their DHCP or addressing scheme to use the same IP address). We should be able to add new ones, and remove old.
> 

Furthermore, the change in OVN DB IP isn't going to propagate to all of the worker nodes in K8s cluster. They would still have the old set of IPs. We used to have watchers to watch for changes to ovnkube-db service, but we don't have it anymore. So, all the worker nodes will not get updated with the new set of IPs.

I don't know of a case where deployers would use DHCP for their K8s nodes. I always thought that the IPs has to be fixed for sane operation of the k8s cluster?

trozet commented:
> > > * also in real deployment, we should pin the OVN DB Pods to a set of K8s nodes so that the OVN DB IPs are same. this is akin to how etcd servers will be managed for K8s cluster
> > 
> > 
> > I don't think that is realistic. A customer could want to add another master node temporarily while he brings down his other one for a hardware upgrade (without having to change their DHCP or addressing scheme to use the same IP address). We should be able to add new ones, and remove old.
> 
> Furthermore, the change in OVN DB IP isn't going to propagate to all of the worker nodes in K8s cluster. They would still have the old set of IPs. We used to have watchers to watch for changes to ovnkube-db service, but we don't have it anymore. So, all the worker nodes will not get updated with the new set of IPs.
> 
> I don't know of a case where deployers would use DHCP for their K8s nodes. I always thought that the IPs has to be fixed for sane operation of the k8s cluster?

OCP by default uses DHCP for the entire cluster. A network admin would typically have static DHCP entries based on mac address to ensure the nodes always got the same IP.

girishmg commented on dist/images/ovndb-raft-functions.sh:
```diff
@@ -17,24 +17,59 @@ verify-ovsdb-raft() {
+  echo "Found ${pod} ip: $init_ip"
+  init_ip=$(bracketify $init_ip)
+  target=$(ovn-${db}ctl --timeout=5 --db=${transport}:${init_ip}:${port} ${ovndb_ctl_ssl_opts} --data=bare --no-headings \
+    --columns=target list connection)
```
L36: I don't think this will work unless you use `--no-leader-only`. Please see below:

```
# ovn-nbctl --timeout=5 --db=tcp:10.0.16.14:6641 --data=bare --no-headings     --columns=target list connection
ovn-nbctl: tcp:10.0.16.14:6641: database connection failed ()
# ovn-nbctl --no-leader-only --timeout=5 --db=tcp:10.0.16.14:6641 --data=bare --no-headings     --columns=target list connection
ptcp:6641
# 
```

girishmg commented on dist/images/ovndb-raft-functions.sh:
```diff
@@ -272,12 +335,14 @@ ovsdb-raft() {
+      if [[ ${db} == "nb" ]]; then
+        set_northd_probe_interval
+      fi
     fi
```
Why are we wrapping the calls to change election_timer and northd_probe_interval under initial_raft_create? I think it's ok to just let it be called all the time since if the values have not changed, then the call is no-op.

girishmg commented on dist/images/ovndb-raft-functions.sh:
```diff
@@ -17,24 +17,59 @@ verify-ovsdb-raft() {
+  for db_pod in $db_pods; do
+    if db_part_of_cluster $db_pod $db $port; then
+      echo "${db_pod} is part of current cluster with ip: ${init_ip}!"
+      return 0
```
L60: this would mean that you could return out of the `for` loop for ovnkube-db-2 and that init_ip set ovkube-db-2 IP address

girishmg commented on dist/images/ovndb-raft-functions.sh:
```diff
@@ -17,24 +17,59 @@ verify-ovsdb-raft() {
   local port=${2}
 
+  db_pods=$(kubectl --server=${K8S_APISERVER} --token=${k8s_token} --certificate-authority=${K8S_CACERT} \
+    get pod -n ${ovn_kubernetes_namespace} -o=jsonpath='{.items[*].metadata.name}' | egrep -o 'ovnkube-db[^ ]+')
```
should we use a label selector here? perhaps `-l name=ovnkube-db`?

trozet commented:
superseded by #1717
